storage/fanout.go

```diff
@@ -338,11 +379,28 @@ func (c *mergeSeriesSet) Next() bool {
 	c.currentLabels = c.heap[0].At().Labels()
 	for len(c.heap) > 0 && labels.Equal(c.currentLabels, c.heap[0].At().Labels()) {
 		set := heap.Pop(&c.heap).(SeriesSet)
+		if remoteSeriesSet, ok := set.(RemoteSet); ok {
+			if contains(c.storage.failedRemoteQueriers, remoteSeriesSet.GetQuerier()) {
+				continue
+			}
+		}
 		c.currentSets = append(c.currentSets, set)
 	}
+	if len(c.currentSets) == 0 {
+		return c.Next()
```

Comment:
Why would that cause a deep recursion? Or any recursion for that matter? If the first query succeeded, then the series would get popped off the heap and added to currentSets, and it wouldn't even recursively call next then because the length would no longer be 0.

Reply:
Unless I misunderstood. Do you mean 2 queries for different remote storages on the same set of label matchers? Or 2 queries for the same remote storage on a different set of label matchers? I assume the former before. If it's the latter, then it still shouldn't cause a deep recursion because the seriesSet was popped off the heap, so the next time that Next() gets called, the series set isn't in the heap or the currentSets, so it wouldn't try to check it again.

Reply:
This line is calling itself, and I'm not really clear on what this logic is doing.

Reply:
The idea is that if for the current labels, all the series sets came from failed remote storages, then you'd end up with an empty currentSets, so you wouldn't want to return true. But it wouldn't want to return false either because the next labels pulled from the heap could have sets from non-failed remote sources (or the primary source). So it should check Next(). If it returned false then subsequent sets would be missed.

Reply:
This is still a recursion, an iterative version would be safer. This logic would also benefit from comments.